network/src/peer_store/sqlite/peer_store.rs

```diff
@@ -68,48 +67,52 @@ impl SqlitePeerStore {
-        for (ip, ban_time) in ban_records {
-            self.ban_list.insert(ip, ban_time);
-        }
+    fn init_banned_addresses(&mut self) -> Result<(), DBError> {
+        db::clear_expires_banned_addresses(&self.conn)?;
+        self.banned_addresses = db::get_banned_addresses(&self.conn)?
+            .iter()
+            .map(|banned_address| (banned_address.address, banned_address.ban_until))
+            .collect();
         Ok(())
     }
 
-    fn ban_ip(&mut self, addr: &Multiaddr, timeout: Duration) {
-        let ip = {
-            match addr.extract_ip_addr_binary() {
-                Some(binary) => binary,
-                None => return,
-            }
-        };
-        let ban_time = unix_time_as_millis() + (timeout.as_millis() as u64);
-        db::insert_ban_record(&self.conn, &ip, ban_time).expect("ban ip");
-        self.ban_list.insert(ip, ban_time);
-        if self.ban_list.len() > BAN_LIST_CLEAR_EXPIRES_SIZE {
-            self.clear_expires_banned_ip().expect("clear ban list");
+    fn ban_ip(&mut self, addr: &Multiaddr, duration: Duration) {
+        if let Some(ip_network) = multiaddr_to_ip_network(addr) {
+            let now = unix_time_as_millis();
+            let ban_until = now + (duration.as_millis() as u64);
+            let banned_address = BannedAddress {
+                address: ip_network,
+                ban_reason: "".to_owned(),
+                ban_until,
+                created_at: now,
+            };
+            self.ban_address(banned_address);
         }
     }
 
+    fn ban_address(&mut self, banned_address: BannedAddress) {
+        db::insert_ban(&self.conn, &banned_address).expect("ban address");
+        self.banned_addresses
+            .insert(banned_address.address, banned_address.ban_until);
+        self.clear_expires_banned_addresses()
+            .expect("clear ban list");
+    }
+
     fn is_addr_banned(&self, addr: &Multiaddr) -> bool {
-        let ip = match addr.extract_ip_addr_binary() {
-            Some(ip) => ip,
-            None => return false,
-        };
-        let now = unix_time_as_millis();
-        match self.ban_list.get(&ip) {
-            Some(ban_time) => *ban_time > now,
-            None => false,
+        if let Some(ip) = addr.extract_ip_addr() {
+            let now = unix_time_as_millis();
+            return self
+                .banned_addresses
+                .iter()
+                .any(|(ip_network, ban_until)| ban_until.gt(&now) && ip_network.contains(ip));
         }
+        false
     }
 
-    fn clear_expires_banned_ip(&mut self) -> Result<(), DBError> {
-        let now = unix_time_as_millis();
-        let ips = db::clear_expires_banned_ip(&self.conn, now)?;
-        for ip in ips {
-            self.ban_list.remove(&ip);
+    fn clear_expires_banned_addresses(&mut self) -> Result<(), DBError> {
+        let addresses = db::clear_expires_banned_addresses(&self.conn)?;
```

Comment:
Don't need to get the expired addresses from the database. It is easy to retain only the effective ban in the memory.